src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+		pdata->res_idx_to_dev_idx[i] = i;
+
+		ratbag_resolution_set_cap(resolution,
+					  RATBAG_RESOLUTION_CAP_SEPARATE_XY_RESOLUTION);
+	}
+}
+
+static int
+gskill_get_firmware_version(struct ratbag_device *device) {
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x08 };
+	int rc;
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to read the firmware version of the mouse: %d\n",
+			  rc);
+		return rc;
+	}
+
+	return buf[4];
+}
+
+static unsigned int
+gskill_misc_macro_code_to_keycode(uint8_t code)
+{
+	unsigned int keycode;
+
+	switch (code & 0x0f) {
+	case 0x0: keycode = KEY_LEFTCTRL;   break;
+	case 0x1: keycode = KEY_LEFTSHIFT;  break;
+	case 0x2: keycode = KEY_LEFTALT;    break;
+	case 0x3: keycode = KEY_LEFTMETA;   break;
+	case 0x4: keycode = KEY_RIGHTCTRL;  break;
+	case 0x5: keycode = KEY_RIGHTSHIFT; break;
+	case 0x6: keycode = KEY_RIGHTALT;   break;
+	case 0x7: keycode = KEY_RIGHTMETA;  break;
+	case 0x8: keycode = BTN_LEFT;       break;
+	case 0x9: keycode = BTN_RIGHT;      break;
+	case 0xa: keycode = BTN_MIDDLE;     break;
+	case 0xb: keycode = BTN_SIDE;       break;
+	case 0xc: keycode = BTN_EXTRA;      break;
+	default: keycode = 0;               break;
+	}
+
+	return keycode;
+}
+
+static uint8_t
+gskill_macro_code_from_event(struct ratbag_device *device,
+			     struct ratbag_macro_event *event)
+{
+	uint8_t macro_code;
+
+	/*
+	 * The miscellanious keycodes are ORd with 0x70 to indicate press, 0xF0
+	 * to indicate release
+	 */
+	if (event->type == RATBAG_MACRO_EVENT_KEY_PRESSED)
+		macro_code = 0x70;
+	else
+		macro_code = 0xF0;
+
+	switch (event->event.key) {
+	case KEY_LEFTCTRL:   macro_code |= 0x00; break;
+	case KEY_LEFTSHIFT:  macro_code |= 0x01; break;
+	case KEY_LEFTALT:    macro_code |= 0x02; break;
+	case KEY_LEFTMETA:   macro_code |= 0x03; break;
+	case KEY_RIGHTCTRL:  macro_code |= 0x04; break;
+	case KEY_RIGHTSHIFT: macro_code |= 0x05; break;
+	case KEY_RIGHTALT:   macro_code |= 0x06; break;
+	case KEY_RIGHTMETA:  macro_code |= 0x07; break;
+	case BTN_LEFT:       macro_code |= 0x08; break;
+	case BTN_RIGHT:      macro_code |= 0x09; break;
+	case BTN_MIDDLE:     macro_code |= 0x0a; break;
+	case BTN_SIDE:       macro_code |= 0x0b; break;
+	case BTN_EXTRA:      macro_code |= 0x0c; break;
+	case KEY_SCROLLDOWN: macro_code  = 0x7e; break;
+	case KEY_SCROLLUP:   macro_code  = 0xfe; break;
+	default:
+		macro_code = ratbag_hidraw_get_keyboard_usage_from_keycode(
+		    device, event->event.key);
+
+		if (event->type == RATBAG_MACRO_EVENT_KEY_RELEASED)
+			macro_code += 0x80;
+		break;
+	}
+
+	return macro_code;
+}
+
+/* TODO: judging from how profile names are handled, it's also very likely that
+ * the encoding of macro names isn't actually in UTF-8 (profile names seem to
+ * be encoded like "t\0h\0i\0s\0"). While this will work if ratbag is the only
+ * program reading/writing from the mouse's profiles, this may intimidate
+ * G.Skill's configuration program
```

Comment:
Please stick to the expected behavior under Windows, and do the conversion here.